File: arq/protocols.java

import java.lang.*;
import java.util.Scanner;

public class protocols {

    public static int totalframe=0, frsize=0, cnt = 0;
    public static Scanner sc = new Scanner(System.in);

    public static void main(String args[]) {

        //int totalframe, frsize, cnt = 0;
        

        System.out.print("\nEnter the Total number of frames : ");
        totalframe = sc.nextInt();
        //totalframe+=2;
        System.out.print("Enter the Window Size : ");
        frsize = sc.nextInt();
        //frsize+=2;
        int ch;
        do {
            System.out.print("\n\t\t1.Go back N\n\t\t2.Selective repeat\n\t\t3.Exit\nEnter your choice  :\t");
            ch = sc.nextInt();
            switch (ch) {
                case 1: {
                    Go_back_n();
                    break;
                }
                case 2: {
                    Selective_Repeat();
                    break;
                }
                case 3: {
                    System.out.println("Exitting........");
                    break;
                }

            }
        } while (ch != 3);
        sc.close();
    }

    static void Go_back_n(){

              int[] send = new int[totalframe*10];
            	int[] recieve = new int[totalframe*10];
            	int[] retransmitted = new int[totalframe*10];
            	int i=0;
 				 for(int m=1;m<=totalframe;m++)
 				 {
  					retransmitted[m]=-1;
				  }
                  while (i <= totalframe) {
                    int z = 0;
                    for (int k = i; k < i + frsize && k <= totalframe; k++) {
                        System.out.println("Sending Frame " + k + "...");     
                        send[k]=k;
                    }
                    for (int k = i;k <= totalframe; k++) {
                      
                      char ch;
                      System.out.print("\nDo you want to send the Acknowledgment for frame " + k + " (y/n): ");
                      ch = sc.next().charAt(0);
                     
                      if (ch=='y') {
                        System.out.println("Positive Acknowledgment for Frame " + k + "...");
                        recieve[k]=k;
                        if((k+frsize)<=totalframe)
                        {
                        	System.out.println("Sending Frame  " + (k+frsize) +"...");
                        	send[k+frsize]=k+frsize;
                        }
                        z++;
                      } else {
                        System.out.println("Negative Acknowledgement for Frame Number : " + k );
                        for(int m=k; m < k+frsize && m<=totalframe;m++)
                        {
                        	System.out.println("Retransmitting Frame " + m);
                        	retransmitted[m]=m;
                        }
                        
                        k=k-1;
                      }
                    }
                    System.out.println(); 
                    i = i + z;
                  }
                  System.out.println("\nThe frames that are sended :");
                  for(int m=1;m<=totalframe;m++)
                  {
                    System.out.println(send[m] + " ");
                  }
                  
                  System.out.println("\nThe frames that are recieved are :");
                  for(int m=1;m<=totalframe;m++)
                  {
                    System.out.println(recieve[m] + " ");
                  }
                  
                  System.out.println("\nThe frames that are Retransmitted are :");
                  for(int m=1;m<=totalframe;m++)
                  {
                    if(retransmitted[m] == -1){
             		
                    }else{
                      System.out.println(retransmitted[m] + " ");
                    }
                  }

    }

    static void Selective_Repeat(){

              int[] send = new int[totalframe*10];
              int[] recieve = new int[totalframe*10];
              int[] retransmitted = new int[totalframe*10];
            	int i=0;
              for(int m=1;m<=totalframe;m++)
              {
                  retransmitted[m]=-1;
                }
                 while (i <= totalframe) {
                      int z = 0;
                      for (int k = i; k < i + frsize && k <= totalframe; k++) {
                        System.out.println("Sending Frame " + k + "..." );     
                        send[k]=k;
                      }
                      for (int k = i;k <= totalframe; k++) {
            
                          char ch;
                          System.out.print("\nDo you want to send the Acknowledgment for Frame "+ k + " (y/n): ");
                          ch = sc.next().charAt(0);
              
                          if (ch=='y') {
                                System.out.println("Positive Acknowledgment for Frame "+ k + "...");
                                    recieve[k]=k;
                                if((k+frsize)<=totalframe)
                                {
                                    System.out.println("Sending Frame  " + (k+frsize) + "...");
                                  send[k+frsize]=k+frsize;
                                }
                                z++;
                          } else {
                                System.out.println("Negative Acknowledgement for Frame Number : "+ k );
                                System.out.println("Retransmitting Frame " + k );
                                retransmitted[k]=k;
                                k=k-1;
                          }
                      }
                      System.out.println();
                      i = i + z;
             }
             System.out.println("\nThe frames that are sended :");
             for(int m=1;m<=totalframe;m++)
             {
                System.out.println(send[m] + " ");
            }
  
                System.out.println("\nThe frames that are recieved are :");
            for(int m=1;m<=totalframe;m++)
             {
            	System.out.println(recieve[m] + " ");
             }
  
             System.out.println("\nThe frames that are Retransmitted are :");
             for(int m=1;m<=totalframe;m++)
             {
             	if(retransmitted[m] == -1){
             		
                }else{
                  System.out.println(retransmitted[m] + " ");
                }
            }

    }

}

File: arq/protocols.cpp
#include<iostream>
using namespace std;

class protocols{
	
 public :
            void Selective_Repeat(int i, int frsize, int totalframe, int cnt) {

            	int send[totalframe];
            	int recieve[totalframe];
            	int retransmitted[totalframe];
            	
              for(int m=1;m<=totalframe;m++)
              {
                  retransmitted[m]=-1;
                }
                 while (i <= totalframe) {
                      int z = 0;
                      for (int k = i; k < i + frsize && k <= totalframe; k++) {
                          cout << "Sending Frame " << k << "..." << endl;     send[k]=k;
                      }
                      for (int k = i;k <= totalframe; k++) {
            
                          char ch;
                          cout<<"\nDo you want to send the Acknowledgment for Frame "<<k<<" (y/n): ";
                          cin>>ch;
              
                          if (ch=='y') {
                                    cout << "Positive Acknowledgment for Frame " << k << "..." << endl;
                                    recieve[k]=k;
                                if((k+frsize)<=totalframe)
                                {
                                  cout<<"Sending Frame  "<<k+frsize<<"..."<<endl;
                                  send[k+frsize]=k+frsize;
                                }
                                z++;
                          } else {
                              cout << "Negative Acknowledgement for Frame Number : " << k <<endl;
                              cout << "Retransmitting Frame "<<k<< endl;
                              retransmitted[k]=k;
                              k=k-1;
                          }
                      }
                      cout << "\n";
                      i = i + z;
             }
            cout<<"\nThe frames that are sended :";
             for(int m=1;m<=totalframe;m++)
             {
  	            cout<<send[m]<<" ";
            }
  
                cout<<"\nThe frames that are recieved are :";
            for(int m=1;m<=totalframe;m++)
             {
            	cout<<recieve[m]<<" ";
             }
  
             cout<<"\nThe frames that are Retransmitted are :";
             for(int m=1;m<=totalframe;m++)
             {
             	if(retransmitted[m] == -1){
             		
				}else{
					cout<<retransmitted[m]<<" ";
				}
             }
  
        }
        
        void Go_back_n(int i, int frsize, int totalframe, int cnt) {
            	int send[totalframe];
            	int recieve[totalframe];
            	int retransmitted[totalframe];
            	
 				 for(int m=1;m<=totalframe;m++)
 				 {
  					retransmitted[m]=-1;
				  }
                  while (i <= totalframe) {
                    int z = 0;
                    for (int k = i; k < i + frsize && k <= totalframe; k++) {
                      cout << "Sending Frame " << k << "..." << endl;     send[k]=k;
                    }
                    for (int k = i;k <= totalframe; k++) {
                      
                      char ch;
                      cout<<"\nDo you want to send the Acknowledgment for frame "<<k<<" (y/n): ";
                      cin>>ch;
                     
                      if (ch=='y') {
                        cout << "Positive Acknowledgment for Frame " << k << "..." << endl;
                        recieve[k]=k;
                        if((k+frsize)<=totalframe)
                        {
                        	cout<<"Sending Frame  "<<k+frsize<<"..."<<endl;
                        	send[k+frsize]=k+frsize;
                        }
                        z++;
                      } else {
                        cout << "Negative Acknowledgement for Frame Number : " << k <<endl;
                        for(int m=k; m < k+frsize && m<=totalframe;m++)
                        {
                        	cout << "Retransmitting Frame "<<m<< endl;
                        	retransmitted[m]=m;
                        }
                        
                        k=k-1;
                      }
                    }
                    cout << "\n";
                    i = i + z;
                  }
                  cout<<"\nThe frames that are sended :";
                  for(int m=1;m<=totalframe;m++)
                  {
                  	cout<<send[m]<<" ";
                  }
                  
                  cout<<"\nThe frames that are recieved are :";
                  for(int m=1;m<=totalframe;m++)
                  {
                  	cout<<recieve[m]<<" ";
                  }
                  
                  cout<<"\nThe frames that are Retransmitted are :";
                  for(int m=1;m<=totalframe;m++)
                  {
                  	cout<<retransmitted[m]<<" ";
                  }
  
            }


};

int main() {
  protocols gbn;
  int totalframe, frsize, cnt = 0;
 
  cout << "Enter the Total number of frames : ";
  cin >> totalframe;
  cout << "Enter the Window Size : ";
  cin >> frsize;
  int i = 1,ch;
  do{
      cout<<"\n1.Go back N\n2.SElective repeat\n3.Exit\nEnter your choice :"<<endl;
      cin>>ch;
      switch(ch)
      {
          case 1:{
               gbn.Go_back_n(i, frsize, totalframe, cnt);
               break;
          }
          
          case 2:
              {
                gbn.Selective_Repeat(i, frsize, totalframe, cnt);  
                 break;
              }
          case 3:
              {
                  cout<<"Exitting....";
                  break;
              }
      
        
        
      }
      }while (ch!=3);
  return 0;
}
